### ISSUE TITLE
FasterRCNN: `TypeError: target labels must of int64 type, instead got torch.int32`

### ISSUE THREAD

**Issue by Kpraetori** [DeepLabCut/DeepLabCut#2642]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

I'm trying to do transfer learning with the Superanimal Quadruped model but I keep getting a couple errors and it won't train. Maybe it's an install issue but I'm not sure and don't know how to resolve it.

1. it throws a notice that I set a batch size of 1 or freeze bn stats = false, neither of which is accurate according to both config files. The config files are attached as text files (won't let me upload .yaml).
2. When it begins start object detector training, a series of traceback errors are thrown and it does not train. I'm attaching the log output.

I have tried on two different servers and rebuilt the environment several times. I have verified nvidia-smi shows the GPU and that torch.cuda.is_available shows true in the environment. I have tried with the latest version of DLC. I have also tried with Pytorch 12.1 and 12.4 in case a newer version was needed. I have followed the regular installation guide in the docs and also the instructions at https://github.com/DeepLabCut/DeepLabCut/pull/2613 to no avail.

I am coming up to a submission deadline so help is really appreciated!

[Complete_Log_DLC_Failing.txt](https://github.com/user-attachments/files/16034320/Complete_Log_DLC_Failing.txt)
[config.txt](https://github.com/user-attachments/files/16034384/config.txt)
[pose_cfg.txt](https://github.com/user-attachments/files/16034385/pose_cfg.txt)
[pytorch_config.txt](https://github.com/user-attachments/files/16034386/pytorch_config.txt)

### Operating System

Windows 2022 Server

### DeepLabCut version

deeplabcut-3.0.0rc1

### DeepLabCut mode

single animal

### Device type

gpu, NVIDIA RTX A4000. I also tried a different server with NVIDIA A10.

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 537.42                 Driver Version: 537.42       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA RTX A4000             WDDM  | 00000000:21:00.0 Off |                  Off |
| 41%   30C    P8               9W / 140W |    508MiB / 16376MiB |      1%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      4872    C+G   C:\Windows\explorer.exe                   N/A      |
|    0   N/A  N/A      8696    C+G   ...cal\Microsoft\OneDrive\OneDrive.exe    N/A      |
|    0   N/A  N/A      9176    C+G   ...m Files\Mozilla Firefox\firefox.exe    N/A      |
|    0   N/A  N/A     10988    C+G   ...crosoft\Edge\Application\msedge.exe    N/A      |
|    0   N/A  N/A     11060    C+G   ...CBS_cw5n1h2txyewy\TextInputHost.exe    N/A      |
|    0   N/A  N/A     12368    C+G   ...ekyb3d8bbwe\PhoneExperienceHost.exe    N/A      |
|    0   N/A  N/A     17328    C+G   ....Search_cw5n1h2txyewy\SearchApp.exe    N/A      |
+---------------------------------------------------------------------------------------+
```

### Steps To Reproduce

- follow the instructions at: https://github.com/DeepLabCut/DeepLabCut/pull/2613 for installation
- Open GUI
- Split dataset 80/20
- Then go to train network. It will return the log I am adding below. The full log output is attached in the bug description.

### Relevant log output

```shell
Note: According to your model configuration, you're training with batch size 1 and/or ``freeze_bn_stats=false``. This is not an optimal setting if you have powerful GPUs.
This is good for small batch sizes (e.g., when training on a CPU), where you should keep ``freeze_bn_stats=true``.
If you're using a GPU to train, you can obtain faster performance by setting a larger batch size (the biggest power of 2 where you don't geta CUDA out-of-memory error, such as 8, 16, 32 or 64 depending on the model, size of your images, and GPU memory) and ``freeze_bn_stats=false`` for the backbone of your model.
This also allows you to increase the learning rate (empirically you can scale the learning rate by sqrt(batch_size) times).

Using 628 images and 157 for testing

Starting object detector training...
--------------------------------------------------
Traceback (most recent call last):
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\gui\tabs\train_network.py", line 190, in train_network
    compat.train_network(config, shuffle, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\compat.py", line 245, in train_network
    return train_network(
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\apis\train.py", line 326, in train_network
    train(
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\apis\train.py", line 189, in train
    runner.fit(
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\runners\train.py", line 170, in fit
    train_loss = self._epoch(
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\runners\train.py", line 221, in _epoch
    losses_dict = self.step(batch, mode)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\runners\train.py", line 503, in step
    losses, predictions = self.model(images, target)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\models\detectors\fasterRCNN.py", line 106, in forward
    return self.model(x, targets)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torchvision\models\detection\generalized_rcnn.py", line 105, in forward
    detections, detector_losses = self.roi_heads(features, proposals, images.image_sizes, targets)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torch\nn\modules\module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torch\nn\modules\module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\torchvision\models\detection\roi_heads.py", line 749, in forward
    raise TypeError(f"target labels must of int64 type, instead got {t['labels'].dtype}")
TypeError: target labels must of int64 type, instead got torch.int32
```

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by n-poulsen** [comment]
@Kpraetori this looks like an issue where the wrong data type is passed to the detector (`int32` instead of `int64`) - I'll work on replicating it and get back to you

**Comment by Kpraetori** [comment]
Wonderful, thank you! If I can help in any way I'm more than happy to do so.

**Comment by ayayagoesbananas** [comment]
Hello
I actually have the same problem. Here's the bug that I got :
 **' raise TypeError(f"target labels must of int64 type, instead got {t['labels'].dtype}")
TypeError: target labels must of int64 type, instead got torch.int32'**
How can I fix it? Thank you.

**Comment by fmachta** [comment]
Yes, I am experiencing the same issue as well.

    raise TypeError(f"target labels must of int64 type, instead got {t['labels'].dtype}")
TypeError: target labels must of int64 type, instead got torch.int32

Is there a solution yet? thanks

**Comment by SowonKIMM** [comment]
I use two animals. At the  "Create training data set", I set the "Weight Initialziation" as "Transfer Learning-Image Net"  ,  "Network architechture" as "animaltokenpose_base" and the Augmentation method as albumentations. 

I also get totaly same error at the fasterRCNN.py in forward function.

I guess the get_target function of Faster RCNN makes the problem. As the output variable res contains 'labels' and 'image_id'. The tensor type of those values are int 32, so I cast the type by myself. 

I recommends you to go to path and open the file fasterRCNN.py 
 "C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\models\detectors\fasterRCNN.py", 
![KakaoTalk_20240709_005010255](https://github.com/DeepLabCut/DeepLabCut/assets/175037297/fe56945b-9cc4-4b13-a2cd-00903e44ef90)

And change the line 159-166 as the figure 
I hope it works well 
@n-poulsen  please check the solution

**Comment by zclecle2** [comment]
After changing the code and restart the deeplabcut, it solves the bugs. Thank you for your useful suggestions! @SowonKIMM

**Comment by Kpraetori** [comment]
@SowonKIMM thanks, that worked for fixing that bug!

Unfortunately other issues are cropping up still. So because of debugging to try and ID the issue, I wanted to start with a fresh environment in case I made accidental changes. I followed the instructions as listed above, but this time the GUI wouldn't load. I went through and tried the install instructions from [Issue 2644](https://github.com/DeepLabCut/DeepLabCut/issues/2644) and after a few re-attempts got this to work:

```
conda create -n deeplabcut3 python=3.11
conda activate deeplabcut3
conda install pytorch torchvision pytorch-cuda=12.2 -c pytorch -c nvidia
conda install -c conda-forge pytables==3.8.0
pip install "git+https://github.com/DeepLabCut/DeepLabCut.git@pytorch_dlc#egg=deeplabcut[gui,modelzoo,wandb]"
Pip install numexpr==2.8.4
Pip install pillow==10.4.0
```

I changed the code in fasterRCNN.py before proceeding.

With the new env, I got torch.cuda.is_available to show as true. I started training and it still showed the batch error of 1 and was horribly slow, over in a day and a half had not even hit 50 epochs. We had to take the server down for something so I stopped training. I checked and nvcc -v and it did not show as working. 

I'm not sure what I have to install or do to make nvcc show as true. Is there a dependency that's not installing as it should?

**Comment by recant9226** [comment]
Hi @SowonKIMM and @n-poulsen , the change in this thread helped fix the second issue that I was having similar to @Kpraetori. i.e. the type change to "torch.int64". 

I also have the same problem as @Kpraetori (problem #1) where the training is now slow similar to Kpraetori's complaint: "it throws a notice that I set a batch size of 1 or freeze bn stats = false, neither of which is accurate according to both config files. The config files are attached as text files (won't let me upload .yaml)." My config files reflect the text files that @Kpraetori uploaded, but I noticed the "fasterRCNN.py" file has a "line freeze_bn_stats: bool = False,". Should I try switching this to "freeze_bn_stats: bool = True," or was there already a solution found to fix this issue?

Please let me know at your own convenience. Also, let me know if it is appropriate to continue this thread or if I should instead start a new thread.

Cheers!

**Comment by Kpraetori** [comment]
@recant9226 So I made some edits and this time didn't get the error:

- Make a new env using the latest version, deeplabcut-3.0.0rc1
     - I had to fiddle with the order of what to install when setting up the env but this is what worked for me and got nvcc -V to show a result and torch is available to also show true:
<code>
conda create -n deeplabcut3 python=3.11
conda activate deeplabcut3
conda install cuda -c nvidia/label/cuda-12.2.0
conda install pytorch torchvision torchaudio pytorch-cuda=12.1 -c pytorch -c nvidia
conda install cudnn -c conda-forge
conda install -c conda-forge pytables==3.8.0
pip install "git+https://github.com/DeepLabCut/DeepLabCut.git@pytorch_dlc#egg=deeplabcut[gui,modelzoo,wandb]" 
</ code>
 
- from `"C:\Other_Program_Files\miniforge3\envs\deeplabcut3\lib\site-packages\deeplabcut\pose_estimation_pytorch\models\detectors\fasterRCNN.py",` changed line 48 (freeze_bin_stats) to True
- From my pytorch_config file I checked that line 48 was True while line 110 was False as per the note instructions. 

I just started training the model so not sure if it's working properly, but at least the error is not showing anymore.

**Comment by n-poulsen** [comment]
This should have been fixed in #2676, where targets are explicitly cast to `long()` before being given to the target generator.
